rfcs/0085-nixos-release-stablization.md

```diff
@@ -0,0 +1,128 @@
+discredit any individual, but demonstrate that certain packages have
+many dimensions of compatibility which need a longer time to stabilize
+on unstable before being included in a release.
+
+
+# Definitions
+[definitions]: #definition
+- Breaking change
+  - A change that is likely to break downstream.
+  - For projects following Semantic Versioning, disallow major version bumps.
+- Restricted
+  - Breaking changes are disallowed.
+  - Breaking changes for security or critical bugs are allowed. (Patching is preferred over updates.)
+  - New packages and NixOS modules are welcome.
+- Unrestricted
+  - Breaking changes are allowed.
+  - New packages and NixOS modules are welcome.
+- ZHF: Zero Hydra Failures
+  - Period in which packages are stabilized in preparation of a release.
+
+# Changes
+
+## Release Critical Packages
+
+A list of Release Critical Packages will be maintained in the nixpkgs
+manual.  Release Critical Packages are important packages which have
+many dimensions of build or runtime behavior. Generally these will be
+packages which are a part of `stdenv`, or bootstrapping a system
+(e.g. systemd). The Release Managers are empowered to decide which
+packages are Release Critical Packages without additional RFCs.
+
+Initially these will be `stdenv.cc`,`binutils`, and `systemd`.
+
+## Restriction Timeline
+
+This timeline represents when certain branches and packages will be
+restricted. It is not a complete timeline of all release activities.
+The actual timeline will still be determined by the Release Managers
+and may be adjusted as needed. The current position in the timeline
+will be communicated on Discourse.
+
+| Weeks from Release | Branches Affected | Events |
+| --- | --- | --- |
+| -8 Weeks | | Gnome and Plasma(YY.11) packaging begins |
+| -6 Weeks | `staging-next`, `staging` | Restrict breaking changes to Release Critical Packages |
+| -4 Weeks | `staging-next`, `staging` | Restrict all breaking changes: allow only non-breaking updates and Desktop Manager changes |
+| -3 Weeks | `master` | (Day before ZHF) merge in `staging-next`, prep for ZHF |
+| -3 Weeks | `master` | Begin ZHF |
+| -3 Weeks | `master` | Focus on minimizing regressions in PRs |
+| -2 Weeks | `master` | Merge first `staging-next` fixes; begin second `staging-next` fix cycle |
+| -2 Weeks | `staging` | Unrestrict all breaking changes; new changes will not be present in `master` before branch-off |
+| -1 Weeks | `master` | Merge second `staging-next` fix cycle |
+| -1 Weeks | `staging-next` | Unrestrict all breaking changes; new changes will not be present in `master` before branch-off |
+| -1 Weeks | `master`, `release` | Perform Branch-off, create release channels, create new beta / unstable tags |
+| -1 Weeks | `master`, `release` | ZHF transitions to "backporting" workflow |
+| -1 Weeks | `release` | Prepare for release, finish remaining issues |
+| 0 Weeks | `release` | Release! |
+| 0 Weeks | | ZHF Ends |
```

Comment:
Did I miss something, or does this not tell when "Restricted" mode begins and ends on `master`?

Reply:
You didn't miss anything. I mention [it above](https://github.com/NixOS/rfcs/pull/85#discussion_r602933101), but PRs which target master have a small enough scope that breaking changes can be more easily assessed, and thus don't bear the same potential to break builds as a staging PR.

>  if the major version bump affects <100 (or 500 in stituations, e.g. python packages) rebuilds, then it's still free to be bumped. All PRs which target master will not be affected by any of these changes.

The intent of this RFC is really to limit the amount of "potential breakages" right before a release. This is why breaking changes to `staging` shouldn't be merged. We constantly have to revisit downstream packages which are broken by such changes, and causes a lot of work which could be delayed til after the branchoff. "Non-breaking" staging changes are more aligned to potentially fixing issues and improving the quality of nixpkgs, so they should be allowed to be merged. Obviously not all packages follow semantic versioning, so changelogs should be reviewed and some downstream packages should be tested before merging. However, these tests would ideally happen on all PRs, however staging is particularly difficult as the magnitude of affected packages are much greater, and it's not feasible to thoroughly test every staging PR.

Since most of the risk should be avoided by not bundling 600+ rushed staging PRs immediately before ZHF, we should also be free to do ZHF on master as we should be progressing to a "more stable" state anyway. Also, the hundreds of backport PRs does pose a significant amount of "busy" work for contributors, and I go as far to say the backport PR process is a hindrance to ZHF participation.

Reply:
Ah, so the claim that
> Breaking changes to Release Critical Packages cannot be merged into master for 8 weeks

in the [current "Drawbacks" section](https://github.com/jonringer/rfcs/blob/e3d4d7d43fd0e3d32362905966f985b1c7d7a78d/rfcs/0085-nixos-release-stablization.md#drawbacks) follows _implicitly_ from the fact that such changes (both breaking _and_ concerning Release Critical Packages) would have to come to `master` _through_ `staging` or `staging-next` anyway rather than through regular PRs towards `master`?

Reply:
Correct, I guess the wording could be clearer, but yes, "critical packages" will most likely always need to go through `staging`. The later breaking change restriction to all `staging` PRs is meant to diminish  

"Critical Package" in this context would be a package which has significant downstream packages, and likely have multiple dimensions of compatibility (E.g. systemd's headers vs systemd's dbus messaging API). The ladder point about compatibility is the major concern, these behaviors are usually hard to test, or create tests for, and thus need to be used by unstable users to find issues and have additional `staging` iterations which can incorporate the fixes.

One example would be during the 20.09 release, where Plasma users could logout or change users because systemd had unintentionally made some login information hidden through its dbus API. Also, this behavior only occurred in nixpkgs due to plasma 5.19 + the most recent version of systemd at the time. Later versions of plasma didn't have this issue as they were developed against a more recent version of systemd. This should also be mitigated by RFC 80, in which our release timeline should be suitable for releasing newer versions of Plasma and Gnome, increasing the likelihood of kernel+systemd+desktopManager compatibility in general.

Reply:
Updated the wording to the "Drawbacks" section